### PR TITLE
Moving docker options to daemon.json

### PR DESCRIPTION
**Pull request by vteratipally**
As per the new docker guidelines about customizing the options
like adding registry-mirrors, moving the options to daemon.json

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
As per the guidelines about the customizing the docker, moving the docker options to the /etc/docker/daemon
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -1453,39 +1453,57 @@ function create-master-etcd-apiserver-auth {
...
+  echo "Assemble docker options"
+  cat <<EOF >/etc/docker/daemon.json
+{
+  "pidfile": "/var/run/docker.pid",
+  "iptables": false,
+  "ip-masq": false,
+EOF
   if [[ "${TEST_CLUSTER:-}" == "true" ]]; then
-    docker_opts+=" --log-level=debug"
+    cat <<EOF >>/etc/docker/daemon.json
```
Comment (karan) [inline cluster/gce/gci/configure-helper.sh:1464]: for DRY, abstract this in a method that can be called from here:

```
if [[ "${XYZ:-}" == "true" ]]; then
  adddockeropt "foo"
fi
```

Comment (vteratipally) [reply cluster/gce/gci/configure-helper.sh]: added a method addockeropt

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -1453,39 +1453,57 @@ function create-master-etcd-apiserver-auth {
 }
 
 function assemble-docker-flags {
-  echo "Assemble docker command line flags"
-  local docker_opts="-p /var/run/docker.pid --iptables=false --ip-masq=false"
+  echo "Assemble docker options"
+  cat <<EOF >/etc/docker/daemon.json
+{
```
[the remark this answers is not in the document]
Comment (vteratipally) [reply cluster/gce/gci/configure-helper.sh:1480]: added

**Comment by vteratipally** [comment]
/test pull-kubernetes-e2e-gce-100-performance
--

**Comment by vteratipally** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
--

**Comment by bsdnet** [comment]
Any update or comments?

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -1453,48 +1453,76 @@ function create-master-etcd-apiserver-auth {
...
+	if [ "$#" -lt 1 ]; then
+	echo "No arguments are passed while adding docker options. Expect one argument"
+	exit 1
+	elif [ "$#" -gt 1 ]; then
+	echo "Only one argument is accepted"
+	exit 1
+	fi
+	# appends the given input to the docker opts file i.e. /etc/docker/daemon.json file
+	if [ -f "$DOCKER_OPTS_FILE" ]; then
+	cat >> "${DOCKER_OPTS_FILE}" <<EOF
```
Comment (SergeyKanzhelev) [inline cluster/gce/gci/configure-helper.sh:1475]: should we remove the option if it already present in the json? So if kubernetes attempts to add the same setting as already configured, the value will be overrriden.

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -1453,48 +1453,76 @@ function create-master-etcd-apiserver-auth {
...
+  $1
+EOF
+	else # if file is not present it creates a new file.
+	cat > "${DOCKER_OPTS_FILE}" <<EOF
+  {
```
Comment (SergeyKanzhelev) [inline cluster/gce/gci/configure-helper.sh:1480]: should $1 be written into the file here?

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (vteratipally) [reply cluster/gce/gci/configure-helper.sh:1478]: that might complicate the places where we can set multiple values like log-opt which has an internal option values.

Comment (vteratipally) [reply cluster/gce/gci/configure-helper.sh:1478]: that might complicate for the options such as log-opt which has internal structure option value.

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (vteratipally) [reply cluster/gce/gci/configure-helper.sh]: actually this is not required. Removed the code as we are creating a file.

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (SergeyKanzhelev) [reply cluster/gce/gci/configure-helper.sh:1478]: I know, but this is the reason to switch to daemon.json, right? We want to make sure settings will not conlict. Will Docker be fine with multiple entries for the same setting in the daemon.json?

**Comment by SergeyKanzhelev** [comment]
@vteratipally once you'd squash commits, I will re-lgtm and un-hold. Looks good

File: cluster/gce/gci/configure-helper.sh [hunk 5]
```diff
@@ -1461,40 +1462,97 @@ function docker-installed {
...
+  $1
+EOF
+	fi
+}
+
+function set_docker_options_non_ubuntu() {
+  # set docker options mtu and storage driver for non-ubuntu
+  # as it is default for ubuntu
+  if ! command -v apt-get >/dev/null 2>&1; then
+    echo "Unable to add docker options. Bailing out..."
```
Comment (SergeyKanzhelev) [inline cluster/gce/gci/configure-helper.sh:1488]: maybe make this message more descriptive - tell it's an Ubuntu and which options will not be added

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -1461,40 +1462,97 @@ function docker-installed {
...
   fi
+
+  echo "setting network bridge"
   if [[ "${NETWORK_PROVIDER:-}" == "kubenet" || "${NETWORK_PROVIDER:-}" == "cni" ]]; then
     # set docker0 cidr to private ip address range to avoid conflict with cbr0 cidr range
-    docker_opts+=" --bip=169.254.123.1/24"
+    addockeropt "\"bip\": \"169.254.123.1/24\","
   else
-    docker_opts+=" --bridge=cbr0"
+   addockeropt "\"bridge\": \"cbr0\","
```
Comment (SergeyKanzhelev) [inline cluster/gce/gci/configure-helper.sh:1531]:
```suggestion
    addockeropt "\"bridge\": \"cbr0\","
```

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -1461,40 +1462,97 @@ function docker-installed {
...
   fi
 
+  echo "setting registry mirror"
+  # TODO move the registry-mirror completely to /etc/docker/daemon.json
```
Comment (SergeyKanzhelev) [inline cluster/gce/gci/configure-helper.sh:1535]: add your GH account and maybe create an issue for this TODO

Comment (SergeyKanzhelev) [reply cluster/gce/gci/configure-helper.sh:1535]:
```suggestion
  # TODO (vteratipally) move the registry-mirror completely to /etc/docker/daemon.json
```

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (bsdnet) [reply cluster/gce/gci/configure-helper.sh:1478]: Looks thing is becoming complicated using today's solution.

I would like to suggest using /etc/docker/daemon.json directly, and just 
copied it to /etc/docker when nodes boot up in the long run.

Writing script to override looks challenge, and also difficult to maintain.

File: cluster/gce/gci/configure-helper.sh (hunk 5 above)
Comment (bsdnet) [reply cluster/gce/gci/configure-helper.sh:1488]: We only have COS and Ubuntu, right

can we do: 
set_docker_options_for_cos instead

**Comment by bsdnet** [comment]
/lgtm for now

In the long run, I preferred to use /etc/docker/daemon.json to override the file during runtime, which is more easy to maintain. However, we need more experiment on the settings done during run-time like TEST_CLUSTER, which may need to overhaul the whole script. 

To lower the risk and solve current issue, I am okay with current PR

**Review by bsdnet — COMMENTED** [review]
/Approve

**Pull request merged**